### PR TITLE
bindinfo: get all table name from ast for binding fuzz matching

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: ref #48875

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/bindinfo/binding_match.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	return &collectTableName{
+		tableNames: make([]*ast.Node, 0, 4),
+	}
+}
+
+// Enter implements Visitor interface.
+func (c *collectTableName) Enter(in ast.Node) (out ast.Node, skipChildren bool) {
+	switch node := in.(type) {
+	case *ast.TableName, *ast.ColumnName:
+		c.tableNames = append(c.tableNames, &node)
```
Comment (qw4990) [inline pkg/bindinfo/binding_match.go:51]: We can return `in, true` here to skip some children.

File: pkg/bindinfo/binding_match.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+package bindinfo
+
+import (
+	"sync"
+
+	"github.com/pingcap/tidb/pkg/parser/ast"
+)
+
+// ExtractTableName gets all table names from ast.Node.
+func ExtractTableName(in ast.Node) []*ast.Node {
```
Comment (qw4990) [inline pkg/bindinfo/binding_match.go:24]: How about renaming it to `CollectTableNames` and renaming the structure below to `tableNameCollector` to make them aligned with each other?

File: pkg/bindinfo/binding_match_test.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+package bindinfo
+
+import (
+	"testing"
+
+	"github.com/pingcap/tidb/pkg/parser"
+	"github.com/stretchr/testify/require"
+)
+
+func TestExtractTableName(t *testing.T) {
```
Comment (qw4990) [inline pkg/bindinfo/binding_match_test.go:24]: Please add more test cases:
```
select * from t
select * from t where a>10
select * from t as tx where tx.a > 10
select * from t1, t2, t3;
select * from t1 where t1.a > (select max(a) from t2);
select * from t1 where t1.a > (select max(a) from t2 where t2.a > (select max(a) from t3));
```

File: pkg/bindinfo/binding_match.go (hunk 1 above)
Comment (hawkingrei) [reply pkg/bindinfo/binding_match.go:52]: Done

File: pkg/bindinfo/binding_match.go (hunk 2 above)
Comment (hawkingrei) [reply pkg/bindinfo/binding_match.go]: Done

File: pkg/bindinfo/binding_match_test.go (hunk 3 above)
Comment (hawkingrei) [reply pkg/bindinfo/binding_match_test.go:39]: All have been added.

File: pkg/bindinfo/binding_match_test.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		{
+			"select * from t1, t2, t3;",
+			[]string{"t1", "t2", "t3"},
+		},
+		{
+			"select * from t1 where t1.a > (select max(a) from t2);",
+			[]string{"t1", "t1", "t2"},
+		},
+		{
+			"select * from t1 where t1.a > (select max(a) from t2 where t2.a > (select max(a) from t3));",
```
Comment (qw4990) [inline pkg/bindinfo/binding_match_test.go:61]: select a, b, c, d ..

File: pkg/bindinfo/binding_match.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+package bindinfo
+
+import (
+	"slices"
+	"sync"
+
+	"github.com/pingcap/tidb/pkg/parser/ast"
+)
+
+// CollectTableNames gets all table names from ast.Node.
```
Comment (qw4990) [inline pkg/bindinfo/binding_match.go:24]: Please add this comment:
```
This function is mainly for binding fuzzy matching.
For example: 
  `select * from t1 where a<1` --> [t1]
  `select * from db1.t1, t2 where a<1` --> [db1.t1, t2]
```

File: pkg/bindinfo/binding_match.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+func (*tableNameCollector) Leave(in ast.Node) (out ast.Node, ok bool) {
+	return in, true
+}
+
+func (c *tableNameCollector) GetResult() []*ast.TableName {
+	return slices.Clone(c.tableNames)
+}
+
+func (c *tableNameCollector) DestroyAndPutToPool() {
+	c.tableNames = c.tableNames[:0]
```
Comment (qw4990) [inline pkg/bindinfo/binding_match.go:67]: Could you implement a standard `reset()` func for `tableNameCollector`

**Review by qw4990 — APPROVED** [review]

File: pkg/bindinfo/binding_match.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+import (
+	"slices"
+	"sync"
+
+	"github.com/pingcap/tidb/pkg/parser/ast"
+)
+
+// CollectTableNames gets all table names from ast.Node.
+func CollectTableNames(in ast.Node) []*ast.TableName {
```
Comment (qw4990) [inline pkg/bindinfo/binding_match.go:25]: And please also comment that the returned `[]*ast.TableName` is read-only.

File: pkg/bindinfo/binding_match_test.go (hunk 4 above)
Comment (hawkingrei) [reply pkg/bindinfo/binding_match_test.go:61]: Done

**Review by elsa0520 — APPROVED** [review]
LGTM

File: pkg/bindinfo/binding_match.go (hunk 5 above)
Comment (hawkingrei) [reply pkg/bindinfo/binding_match.go:23]: updated

File: pkg/bindinfo/binding_match.go (hunk 7 above)
Comment (hawkingrei) [reply pkg/bindinfo/binding_match.go:32]: updated

**Pull request merged**